File: src/py_pkg/py_pkg/param_node.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult


class ParamNode(Node):
    def __init__(self):
        super().__init__("py_param_node")
        self.declare_parameter("my_int", 42)
        self.declare_parameter("my_double", 3.14)
        self.declare_parameter("my_string", "hello")
        self.declare_parameter("my_bool", True)

        self.add_on_set_parameters_callback(self.on_param_change)
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.get_logger().info("Python param node started")

    def on_param_change(self, params):
        for p in params:
            self.get_logger().info(f"Param changed: {p.name} = {p.value}")
        return SetParametersResult(successful=True)

    def timer_callback(self):
        val = self.get_parameter("my_int").value
        self.get_logger().info(f"my_int = {val}")


def main(args=None):
    rclpy.init(args=args)
    node = ParamNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: src/cpp_pkg/src/param_node.cpp
#include "rclcpp/rclcpp.hpp"

class ParamNode : public rclcpp::Node {
public:
    ParamNode() : Node("cpp_param_node") {
        this->declare_parameter("my_int", 42);
        this->declare_parameter("my_double", 3.14);
        this->declare_parameter("my_string", "hello");
        this->declare_parameter("my_bool", true);

        param_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(&ParamNode::on_param_change, this, std::placeholders::_1));

        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&ParamNode::timer_callback, this));

        RCLCPP_INFO(this->get_logger(), "C++ param node started");
    }

private:
    rcl_interfaces::msg::SetParametersResult on_param_change(
        const std::vector<rclcpp::Parameter>& params) {
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        for (const auto& p : params) {
            RCLCPP_INFO(this->get_logger(), "Param changed: %s = %s",
                p.get_name().c_str(), p.value_to_string().c_str());
        }
        return result;
    }

    void timer_callback() {
        auto val = this->get_parameter("my_int").as_int();
        RCLCPP_INFO(this->get_logger(), "my_int = %ld", val);
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ParamNode>());
    rclcpp::shutdown();
    return 0;
}
